File: EXAMENES/Sep16/haskell/Huffman.hs
-- | Data Structures
-- | September, 2016
-- |
-- | Student's name:
-- | Student's group:

module Huffman where

import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.WBLeftistHeapPriorityQueue as PQ
import Data.List (nub)

-- | Exercise 1

-- weights :: Ord a => [a] -> D.Dictionary a Int
-- weights xs = aux (nub xs) xs D.empty
--   where
--     aux [] l d = d
--     aux (x:xs) l d = aux xs l (D.insert x (frecuencia x l) d)
--      where 
--       frecuencia x l = foldr (\cab cola -> if cab == x then 1 + cola else cola) 0 l

weights :: Ord a => [a] -> D.Dictionary a Int
weights xs = aux (nub xs) xs D.empty
  where
    aux [] d = d
    aux (x:xs) d
      |D.isDefinedAt x d = aux xs (D.insert x (fromJust (D.valueOf x d)+1) d)
      |otherwise = aux xs (D.insert x 1 d)

fromJust :: Maybe a -> a
fromJust Nothing = error "Nothing"
fromJust (Just x) = x

{-

> weights "abracadabra"
AVLDictionary('a'->5,'b'->2,'c'->1,'d'->1,'r'->2)

> weights [1,2,9,2,0,1,6,1,5,5,8]
AVLDictionary(0->1,1->3,2->2,5->2,6->1,8->1,9->1)

> weights ""
AVLDictionary()

-}


-- Implementation of Huffman Trees
data WLeafTree a = WLeaf a Int  -- Stored value (type a) and weight (type Int)
                 | WNode (WLeafTree a) (WLeafTree a) Int -- Left child, right child and weight
                 deriving (Eq, Show)

weight :: WLeafTree a -> Int
weight (WLeaf _ n)   = n
weight (WNode _ _ n) = n

-- Define order on trees according to their weights
instance Eq a => Ord (WLeafTree a) where
  wlt <= wlt' =  weight wlt <= weight wlt'

-- Build a new tree by joining two existing trees
merge :: WLeafTree a -> WLeafTree a -> WLeafTree a
merge wlt1 wlt2 = WNode wlt1 wlt2 (weight wlt1 + weight wlt2)


-- | Exercise 2

-- 2.a
huffmanLeaves :: String -> PQ.PQueue (WLeafTree Char)
huffmanLeaves s = aux (nub s) (weights s) (PQ.empty)
  where 
    aux [] w q = q
    aux (x:xs) w q = aux xs w (PQ.enqueue (WLeaf x (fromJust (D.valueOf x w))) q)

{-

> huffmanLeaves "abracadabra"
WBLeftistHeapPriorityQueue(WLeaf 'c' 1,WLeaf 'd' 1,WLeaf 'b' 2,WLeaf 'r' 2,WLeaf 'a' 5)

-}

-- 2.b
huffmanTree :: String -> WLeafTree Char
huffmanTree s = if length (nub s) < 2 then error "La cadena tiene que tener al menos 2 caracteres distintos" else aux (PQ.first (huffmanLeaves s)) (PQ.dequeue (huffmanLeaves s))
  where
    aux h q
      |PQ.isEmpty q = h
      |otherwise = aux (merge h (PQ.first q)) (PQ.dequeue q)

{-

> printWLeafTree $ huffmanTree "abracadabra"
      11_______
      /        \
('a',5)        6_______________
              /                \
        ('r',2)          _______4
                        /       \
                       2        ('b',2)
                      / \
                ('c',1) ('d',1)

> printWLeafTree $ huffmanTree "abracadabra pata de cabra"
                         ______________________25_____________
                        /                                     \
         ______________10______                          ______15
        /                      \                        /       \
       4_______                6_______                6        ('a',9)
      /        \              /        \              / \
('d',2)        2        (' ',3)        3        ('b',3) ('r',3)
              / \                     / \
        ('e',1) ('p',1)         ('t',1) ('c',2)

> printWLeafTree $ huffmanTree "aaa"
*** Exception: huffmanTree: the string must have at least two different symbols

-}


-- | Exercise 3

-- 3.a
joinDics :: Ord a => D.Dictionary a b -> D.Dictionary a b -> D.Dictionary a b
joinDics d1 d2 = aux (D.keysValues d1) d2
  where
    aux [] d2 = d2
    aux (x:xs) d2 = aux xs (D.insert (fst x) (snd x) d2)

{-

> joinDics (D.insert 'a' 1 $ D.insert 'c' 3 $ D.empty) D.empty
AVLDictionary('a'->1,'c'->3)

> joinDics (D.insert 'a' 1 $ D.insert 'c' 3 $ D.empty) (D.insert 'b' 2 $ D.insert 'd' 4 $ D.insert 'e' 5 $ D.empty)
AVLDictionary('a'->1,'b'->2,'c'->3,'d'->4,'e'->5)

-}

-- 3.b
prefixWith :: Ord a => b -> D.Dictionary a [b] -> D.Dictionary a [b]
prefixWith x d = aux (zip (D.keys d) (map (x:) (D.values d))) d
  where
    aux [] d = d
    aux (x:xs) d = aux xs (D.insert (fst x) (snd x) d)

{-

> prefixWith 0 (D.insert 'a' [0,0,1] $ D.insert 'b' [1,0,0] $ D.empty)
AVLDictionary('a'->[0,0,0,1],'b'->[0,1,0,0])

> prefixWith 'h' (D.insert 1 "asta" $ D.insert 2 "echo" $ D.empty)
AVLDictionary(1->"hasta",2->"hecho")

-}

-- 3.c
huffmanCode :: WLeafTree Char -> D.Dictionary Char [Integer]
huffmanCode t = aux t [] D.empty
  where
    aux (WLeaf k v) l d = (D.insert k l d)
    aux (WNode lt rt _) l d = joinDics (aux lt (l++0) d) (aux rt (l++1) d)
    

{-

> huffmanCode (huffmanTree "abracadabra")
AVLDictionary('a'->[0],'b'->[1,1,1],'c'->[1,1,0,0],'d'->[1,1,0,1],'r'->[1,0])

-}

-- ONLY for students not taking continuous assessment

-- | Exercise 4

encode :: String -> D.Dictionary Char [Integer] -> [Integer]
encode = undefined

{-

> encode "abracadabra" (huffmanCode (huffmanTree "abracadabra"))
[0,1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0]

-}

-- | Exercise 5

-- 5.a
takeSymbol :: [Integer] -> WLeafTree Char -> (Char, [Integer])
takeSymbol = undefined

{-

> takeSymbol [0,1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0] (huffmanTree "abracadabra")
('a',[1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0])

> takeSymbol [1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0] (huffmanTree "abracadabra")
('b',[1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0])

-}

-- 5.b
decode :: [Integer] -> WLeafTree Char -> String
decode = undefined

{-

> decode [0,1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0] (huffmanTree "abracadabra")
"abracadabra"

-}

-------------------------------------------------------------------------------
-- Pretty Printing a WLeafTree
-- (adapted from http://stackoverflow.com/questions/1733311/pretty-print-a-tree)
-------------------------------------------------------------------------------

printWLeafTree :: (Show a) => WLeafTree a -> IO ()
printWLeafTree t = putStrLn (unlines xss)
 where
   (xss,_,_,_) = pprint t

pprint :: Show a => WLeafTree a -> ([String], Int, Int, Int)
pprint (WLeaf x we)             =  ([s], ls, 0, ls-1)
  where
    s = show (x,we)
    ls = length s
pprint (WNode lt rt we)         =  (resultLines, w, lw'-swl, totLW+1+swr)
  where
    nSpaces n = replicate n ' '
    nBars n = replicate n '_'
    -- compute info for string of this node's data
    s = show we
    sw = length s
    swl = div sw 2
    swr = div (sw-1) 2
    (lp,lw,_,lc) = pprint lt
    (rp,rw,rc,_) = pprint rt
    -- recurse
    (lw',lb) = if lw==0 then (1," ") else (lw,"/")
    (rw',rb) = if rw==0 then (1," ") else (rw,"\\")
    -- compute full width of this tree
    totLW = maximum [lw', swl,  1]
    totRW = maximum [rw', swr, 1]
    w = totLW + 1 + totRW
{-
A suggestive example:
     dddd | d | dddd__
        / |   |       \
      lll |   |       rr
          |   |      ...
          |   | rrrrrrrrrrr
     ----       ----           swl, swr (left/right string width (of this node) before any padding)
      ---       -----------    lw, rw   (left/right width (of subtree) before any padding)
     ----                      totLW
                -----------    totRW
     ----   -   -----------    w (total width)
-}
    -- get right column info that accounts for left side
    rc2 = totLW + 1 + rc
    -- make left and right tree same height
    llp = length lp
    lrp = length rp
    lp' = if llp < lrp then lp ++ replicate (lrp - llp) "" else lp
    rp' = if lrp < llp then rp ++ replicate (llp - lrp) "" else rp
    -- widen left and right trees if necessary (in case parent node is wider, and also to fix the 'added height')
    lp'' = map (\s -> if length s < totLW then nSpaces (totLW - length s) ++ s else s) lp'
    rp'' = map (\s -> if length s < totRW then s ++ nSpaces (totRW - length s) else s) rp'
    -- first part of line1
    line1 = if swl < lw' - lc - 1 then
                nSpaces (lc + 1) ++ nBars (lw' - lc - swl) ++ s
            else
                nSpaces (totLW - swl) ++ s
    -- line1 right bars
    lline1 = length line1
    line1' = if rc2 > lline1 then
                line1 ++ nBars (rc2 - lline1)
             else
                line1
    -- line1 right padding
    line1'' = line1' ++ nSpaces (w - length line1')
    -- first part of line2
    line2 = nSpaces (totLW - lw' + lc) ++ lb
    -- pad rest of left half
    line2' = line2 ++ nSpaces (totLW - length line2)
    -- add right content
    line2'' = line2' ++ " " ++ nSpaces rc ++ rb
    -- add right padding
    line2''' = line2'' ++ nSpaces (w - length line2'')
    resultLines = line1'' : line2''' : zipWith (\lt rt -> lt ++ " " ++ rt) lp'' rp''

File: ED del Pedro/ExamenesED/Septiembre2016/DataStructures2/java/src/dataStructures/Huffman/Huffman.java
package dataStructures.Huffman;
/**
 * dataStructures.Huffman trees and codes.
 *
 * Data Structures, Grado en Informatica. UMA.
 *
 *
 * Student's name:
 * Student's group:
 */

import dataStructures.dictionary.AVLDictionary;
import dataStructures.dictionary.Dictionary;
import dataStructures.list.LinkedList;
import dataStructures.list.List;
import dataStructures.priorityQueue.BinaryHeapPriorityQueue;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.priorityQueue.WBLeftistHeapPriorityQueue;
import dataStructures.tuple.Tuple2;

public class Huffman {

    // Exercise 1
    public static Dictionary<Character, Integer> weights(String s) {

        Dictionary<Character, Integer> dic = new AVLDictionary<>();

        for (int i = 0; i < s.length(); i ++){
            Character c = s.charAt(i);
            Integer v = dic.valueOf(c);
            if(v == null){
                dic.insert(c, 1);
            }else {
                dic.insert(c, v + 1);
            }
        }

        return dic;
    }

    // Exercise 2.a
    public static PriorityQueue<WLeafTree<Character>> huffmanLeaves(String s) {

        PriorityQueue<WLeafTree<Character>> pq = new WBLeftistHeapPriorityQueue<>();
        Dictionary<Character, Integer> dic = weights(s);

        for (Tuple2<Character, Integer> tup : dic.keysValues()){
            pq.enqueue(new WLeafTree<>(tup._1(), tup._2()));
        }

        return pq;
    }

    // Exercise 2.b
    public static WLeafTree<Character> huffmanTree(String s) {

        PriorityQueue<WLeafTree<Character>> pq = huffmanLeaves(s);

        return huffmanTreeRec(pq.first(), pq);

    }

    private static WLeafTree<Character> huffmanTreeRec(WLeafTree<Character> first, PriorityQueue<WLeafTree<Character>> pq){

        pq.dequeue();

        if(pq.isEmpty()){
            return first;
        }else {
            WLeafTree<Character> wl = pq.first();
            return huffmanTreeRec(merge(first, wl), pq);
        }

    }

    private static WLeafTree<Character> merge(WLeafTree<Character> left, WLeafTree<Character> right){

        return new WLeafTree<>(left, right);

    }

    // Exercise 3.a
    public static Dictionary<Character, List<Integer>> joinDics(Dictionary<Character, List<Integer>> d1, Dictionary<Character, List<Integer>> d2) {

        for (Tuple2<Character, List<Integer>> tup : d1.keysValues()){
            d2.insert(tup._1(), tup._2());
        }

    	return d2;
    }

    // Exercise 3.b
    public static Dictionary<Character, List<Integer>> prefixWith(int i, Dictionary<Character, List<Integer>> d) {

        for (Tuple2<Character, List<Integer>> tup : d.keysValues()){
            tup._2().prepend(i);
        }

    	return d;
    }

    // Exercise 3.c
    public static Dictionary<Character, List<Integer>> huffmanCode(WLeafTree<Character> ht) {

    	return huffmanCodeRec(ht, new AVLDictionary<>(), new AVLDictionary<>());
    }

    private static Dictionary<Character, List<Integer>> huffmanCodeRec(WLeafTree<Character> wlt, Dictionary<Character, List<Integer>> d1, Dictionary<Character, List<Integer>> d2){

        if(wlt.leftChild().isLeaf() && wlt.rightChild().isLeaf()){
            List<Integer> l1 = new LinkedList<>();
            List<Integer> l2 = new LinkedList<>();
            l1.append(0);
            l2.append(1);
            d1.insert(wlt.leftChild().elem(), l1);
            d2.insert(wlt.rightChild().elem(), l2);
            return joinDics(d1, d2);
        }else {
            List<Integer> l2 = new LinkedList<>();
            l2.append(1);
            d2.insert(wlt.rightChild().elem(), l2);
            d1 = prefixWith(0, huffmanCodeRec(wlt.leftChild(), d1, d2));
            return joinDics(d1, d2);
        }
    }

    // Exercise 4
    public static List<Integer> encode(String s, Dictionary<Character, List<Integer>> hc) {
        //to do 
    	return null;
    }

    // Exercise 5
    public static String decode(List<Integer> bits, WLeafTree<Character> ht) {
        //to do 
    	return null;
    }
}
